beacon_node/execution_layer/src/versioned_hashes.rs

```diff
@@ -0,0 +1,130 @@
+extern crate alloy_consensus;
+extern crate alloy_rlp;
+use alloy_consensus::TxEnvelope;
+use alloy_rlp::Decodable;
+use types::{EthSpec, ExecutionPayloadRef, Hash256, Unsigned, VersionedHash};
+
+#[derive(Debug)]
+pub enum Error {
+    DecodingTransaction(String),
+    LengthMismatch { expected: usize, found: usize },
+    VersionHashMismatch { expected: Hash256, found: Hash256 },
+}
+
+pub fn verify_versioned_hashes<E: EthSpec>(
+    execution_payload: ExecutionPayloadRef<E>,
+    expected_versioned_hashes: &[VersionedHash],
+) -> Result<(), Error> {
+    let versioned_hashes =
+        extract_versioned_hashes_from_transactions::<E>(execution_payload.transactions())?;
+    if versioned_hashes.len() != expected_versioned_hashes.len() {
+        return Err(Error::LengthMismatch {
+            expected: expected_versioned_hashes.len(),
+            found: versioned_hashes.len(),
+        });
+    }
+    for (found, expected) in versioned_hashes
+        .iter()
+        .zip(expected_versioned_hashes.iter())
+    {
+        if found != expected {
+            return Err(Error::VersionHashMismatch {
+                expected: *expected,
+                found: *found,
+            });
+        }
+    }
+
+    Ok(())
+}
+
+pub fn extract_versioned_hashes_from_transactions<E: EthSpec>(
+    transactions: &types::Transactions<E>,
+) -> Result<Vec<VersionedHash>, Error> {
+    Ok(transactions
+        .into_iter()
+        .map(beacon_tx_to_tx_envelope)
```

Comment:
this looks like we decode all transactions and then filter them by type. Since the first byte determines the type, can we just filter for blob txs (type 3) first? 

We could probably add a similar irrefutable tx type pattern match during this filter

Reply:
Great point! I've rewritten this to be significantly faster (exiting early on error, only using a single iteration, no need to flatten, etc.)